Fixed #24686 -- Allowed moving a model between apps. (#16905)
django/django: docs/topics/migrations.txt

which reference the field exist. For example, after squashing migrations and
removing the old ones, you should be able to remove the field completely.

.. _moving-model-between-apps:

Moving a model between apps
===========================

> 

Reviewer: ```suggestion

    .. versionadded:: 5.0
```

Django seamlessly handles the process of transferring models from one app

> to another. This involves recognizing the moved model and automatically 
> generating the necessary migrations.  The model's transition occurs within 
> the project state, without affecting the database(unless required). Before 
> running the ``makemigrations`` command, it is mandatory to specify the original
> table name in the Meta section of the model being moved. This prevents the 

Reviewer: ```suggestion
to another. This involves recognizing the moved model and automatically
generating the necessary migrations. The model's transition occurs within
the project state, without affecting the database(unless required). Before
running the ``makemigrations`` command, it is mandatory to specify the original
table name in the Meta section of the model being moved. This prevents the
```

> recreation of the model's table. The process basically involves four steps:

Reviewer: ```suggestion
recreation of the model's table. The process involves four steps:
```

the original table name in the Meta section of the model being moved. This
prevents the recreation of the model's table. The process involves four steps:

* Creating unmanaged moved model in new app.

* Altering model options to make moved model unmanaged in old app.

* Altering model options to make moved model managed in new app.

* Deleting moved model from old app.

> 

Reviewer 2: I think this deserves a more detailed explanation -- in particular, explain *why* we do things, not just *what* we do. Something like:
"""
When we are going to move a model -- that is, delete a model in one app and create it anew in another -- we want to avoid dropping and recreating its database table (which would lose all data). To achieve this, we first create the new model, making it point to the already existing table. To make Django not attempt to create that table (which would fail), we mark it `managed=False` upon creation. After this, we momentarily have two models pointing at the same table. Now, to remove the old model without deleting the table, we mark it `managed=False` as well. Finally, we can remove the old model and mark the new one `managed=True`.
"""

I'm not marking the above as a suggestion, for two reasons:
- It has style issues; in particular, the Django documentation mostly avoids parentheses and m-dashes; and,
- The optional table rename operation is not mentioned (I'm actually not quite sure about its place in the order of operations).

> 

Reviewer: ```suggestion
```

which reference the field exist. For example, after squashing migrations and
removing the old ones, you should be able to remove the field completely.

.. _moving-model-between-apps:

Moving a model between apps
===========================

  .. versionadded:: 5.0

Django seamlessly handles the process of transferring models from one app
to another. This involves recognizing the moved model and automatically
generating the necessary migrations. The model's transition occurs within

> the project state, without affecting the database(unless required). Before

Reviewer 3: ```suggestion
the project state, without affecting the database (unless required). Before
```

I think this whole section might benefit from being the other way around; document that a table rename will happen unless the new model defines a `Meta.db_table` that matches the old one.

running the ``makemigrations`` command, it is mandatory to specify the original
table name in the Meta section of the model being moved. This prevents the
recreation of the model's table. The process involves four steps:

* Creating unmanaged moved model in new app.

* Altering model options to make moved model unmanaged in old app.

* Altering model options to make moved model managed in new app.

* Deleting moved model from old app.

If the moved model is already unmanaged, skip step 2 and 3.

> 

Reviewer 3: Not sure it's worth mentioning this part since it's not typical.

which reference the field exist. For example, after squashing migrations and
removing the old ones, you should be able to remove the field completely.

.. _moving-model-between-apps:

Moving a model between apps
===========================

  .. versionadded:: 5.1

Django seamlessly handles the process of transferring models from one app
to another. This involves recognizing the moved model and automatically
generating the necessary migrations.

When we are going to move a model - that is, delete a model in one app and
create it new in another - we want to avoid dropping and recreating its
database table (which would lose all data). A table rename will happen
unless the moved model in new app defines a ``Meta.db_table`` that matches
the old one.

> To move a model (without a table rename), we first create the new model, making

Reviewer 2: This paragraph explains the migration operations generated by the framework for a model move, and I think this should be said very explicitly, because as the text stands, it can be understood to describe the operations the user needs to take. So I would change the text here along these lines -- clearly separate and describe what the user needs to do, and what the framework does.

Something like:
"""

To move a model from one app to another, move the model's code (the class definition) from the old app's `models.py` file to the new app's `models.py`, and run the `makemigrations` management command. This command will generate migrations in both apps, with the following operations: 
- First, create the new model, making it point to the already existing table. To make Django not attempt to create that table (which would fail), mark it ``managed=False`` upon creation (after this operation, we momentarily have two models pointing at the same table).
- Then, to make it possible to remove the old model without deleting the table, mark it `managed=False` as well.
- Finally, remove the old model and mark the new one `managed=True` (these are two migrations, as each is in its own app)

Note that this only handles the migrations part -- you will still need to handle other aspects of the move (such as changing all existing references to the model to point to its new location) yourself (in fact, you will probably need to handle these before running `makemigrations`, for Django to be able to load the apps).
"""

(the above is my style of writing, Django documentation usually prefers not to use parentheses so liberally)

Author: Thanks 👍 this looks better. Also should i move the last para under a separate `note`?  It'll look something like this: 

<img width="1238" alt="Screenshot 2024-02-06 at 9 02 16 PM" src="https://github.com/django/django/assets/83907321/3ece4b74-c14b-4fd0-9646-cd00abf3511f">

that table (which would fail), we mark it ``managed=False`` upon creation.
After this, we momentarily have two models pointing at the same table. Now,
to remove the old model without deleting the table, we mark it
``managed=False`` as well. Finally, we can remove the old model and mark the
new one ``managed=True``.

To move a model from one app to another, move the model's code (the class definition) from the old app's ``models.py`` file to the new app's ``models.py``, and run the makemigrations management command. This command will generate migrations in both apps, with the following operations:

- First, create the new model, making it point to the already existing table. To make Django not attempt to create that table (which would fail), mark it ``managed=False`` upon creation (after this operation, we momentarily have two models pointing at the same table).
- Then, to make it possible to remove the old model without deleting the table, mark it ``managed=False`` as well.
- Finally, remove the old model and mark the new one ``managed=True`` (these are two migrations, as each is in its own app).

.. note::

>     This only handles the migrations part - you will still need to handle other aspects of the move such as changing all existing references to the model to point to its new location yourself. In fact, you will probably need to handle these before running makemigrations, for Django to be able to load the apps.

Reviewer 4: Hi @DevilsAutumn 👋 excited about this one!

Can you [wrap the documentation at 80 characters](https://docs.djangoproject.com/en/dev/internals/contributing/writing-documentation/#guidelines-for-restructuredtext-files) and update the "makemigrations" references to:
```
:djadmin:`makemigrations`
```
Which will hopefully fix the [spell check errors from the CI](https://github.com/django/django/actions/runs/8418511882/job/23049159127?pr=16905).

Author: @sarahboyce Thanks for the reminder! Fixed the linting errors and rebased as well. 😃

to another. This involves recognizing the moved model and automatically
generating the necessary migrations.

When we are going to move a model - that is, delete a model in one app and
recreate it in another - we want to avoid dropping and recreating its
database table (which would lose all data). A table rename will happen
unless the moved model in the new app defines a ``Meta.db_table`` that matches
the old table name.

To move a model from one app to another, move the model's code (the class
definition) from the old app's ``models.py`` file to the new app's
``models.py``, and run the :djadmin:`makemigrations` management command.
This command will generate migrations in both apps, with the following
operations:

- First, create the new model, making it point to the already existing table. 
  To make Django not attempt to create that table (which would fail), mark it
  ``managed=False`` upon creation (after this operation, we momentarily have
  two models pointing at the same table).
- Then, to make it possible to remove the old model without deleting the
  table, mark it ``managed=False`` as well.
- Finally, remove the old model and mark the new one ``managed=True``

>   (these are two migrations, as each is in its own app).

Reviewer 4: I tried to follow these instructions and must have misinterpreted it and had my test data deleted or I couldn't generate migrations, so I think we might want to spell this out a bit more. 

But specifically

> To move a model from one app to another, move the model's code (the class

This already sounds like the start of instructions, so let's make it visibly part of the instructions of change this e.g 

```
To move a model from one app to another:

#. ...
#. ...
```

The very first paragraph tells you to run the `makemigrations` management command but we don't explicitly say when to migrate or do further makemigrations commands or whether to restrict these commands to a particular app.

We should also add more links to reference docs here (specifically the managed and `db_table` if needed e.g. ```:attr:`Meta.db_table <django.db.models.Options.db_table>` ```)

I think we should also consider giving example names, something like:

```
To move a model `Model1` from `app1` to `app2`:

#. ...
#. ...
```

Then to give an example confusion I had
> To move a model from one app to another, move the model's code (the class
definition) from the old app's ``models.py`` file to the new app's
``models.py``, and run the :djadmin:`makemigrations` management command.

What does "move" mean here? Does move mean "duplicate" (copy and paste) and so you have the model defined in both `app1` and `app2` or does it mean "replace" (cut and paste) and so you only have the model in `app2`? 
I'm hoping have the instructions being like 1. 2. 3. will make this very clear 🤔 

Side note: I'm wondering if we should move the instructions to a 'How to' section and link them here but let's get the instructions really clear first and then decide where they belong after 👍

Author: > What does "move" mean here? Does move mean "duplicate" (copy and paste) and so you have the model defined in both app1 and app2 or does it mean "replace" (cut and paste) and so you only have the model in app2?

"move" actually means moving the model's code from old app to new app(cut and paste). Let me explain with an example:
Suppose you have two apps `app1` and `app2`. In `app1`, you have 2 models `ModelA` and `ModelB` and `ModelA` has a FK to `ModelB`. In `app2`, you have `ModelC`. When you run `makemigrations`, it will create initial migrations for create model.
Now suppose you want to move `ModelB` from `app1` to `app2`, for that you will first cut the code of `ModelB` from `app1` and paste it in `app2`, Now you'll have to resolve the references to `ModelB` (in this case, FK from `ModelA` ) to point to new location of `ModelB` which is `app2.ModelB`. After this you'll just have to run `makemigrations` once and django will automatically generate all the migrations that will be required to move model at database level. Thats it!

>To move a model from one app to another, move the model's code (the class
definition) from the old app's ``models.py`` file to the new app's
``models.py``, and run the :djadmin:`makemigrations` management command.

I think this line is pretty clearly written and will make sense now. ☝️ 

>This command will generate migrations in both apps, with the following
operations:
> - First, create the new model, making it point to the already existing table. 
  To make Django not attempt to create that table (which would fail), mark it
  ``managed=False`` upon creation (after this operation, we momentarily have
  two models pointing at the same table).
> - Then, to make it possible to remove the old model without deleting the
  table, mark it ``managed=False`` as well.
> - Finally, remove the old model and mark the new one ``managed=True``
  (these are two migrations, as each is in its own app).

Now these points are just to explain the migration operations that will be generated automatically and how they will actually "move" a model under the hood. You will gain a better understanding of them once you examine the migrations generated for moving a model.

Maybe its worth adding a minimal example with two apps and few models in the "how-to" section.

which reference the field exist. For example, after squashing migrations and
removing the old ones, you should be able to remove the field completely.

.. _moving-model-between-apps:

Moving a model between apps
===========================

>   .. versionadded:: 5.2

Reviewer 4: ```suggestion
.. versionadded:: 5.2
```

.. note::
    moving inherited models is not yet supported.

> 

Reviewer: OK, the same note is already at the end. Let's remove it.

> .. note::
>     moving inherited models is not yet supported.
> 
> Django seamlessly handles the process of transferring models from one app
> to another. This involves recognizing the moved model and automatically
> generating the necessary migrations.
> 
> When we are going to move a model - that is, delete a model in one app and
> recreate it in another - we want to avoid dropping and recreating its
> database table (which would lose all data). A table rename will happen
> unless the moved model in the new app defines a ``Meta.db_table`` that matches
> the old table name.
> 
> To move a model from one app to another, move the model's code (the class
> definition) from the old app's ``models.py`` file to the new app's
> ``models.py``, and run the :djadmin:`makemigrations` management command.
> This command will generate migrations in both apps, with the following
> operations:
> 
> - First, create the new model, making it point to the already existing table. 
>   To make Django not attempt to create that table (which would fail), mark it
>   ``managed=False`` upon creation (after this operation, we momentarily have
>   two models pointing at the same table).
> - Then, to make it possible to remove the old model without deleting the
>   table, mark it ``managed=False`` as well.
> - Finally, remove the old model and mark the new one ``managed=True``
>   (these are two migrations, as each is in its own app).
> 
> .. note::
>     This only handles the migrations part - you will still need to handle other
>     aspects of the move such as changing all existing references to the model
>     to point to its new location yourself. In fact, you will probably need to
>     handle these before running :djadmin:`makemigrations`, for Django to be able to load
>     the apps.

Reviewer 4: ```suggestion
Django handles the process of transferring models from one app to another. This
involves recognizing the moved model and automatically generating the necessary
migrations.

To move a model ``MyModel`` from ``app1`` to ``app2``:

#. Move the ``MyModel`` class definition from ``app1`` to ``app2``.
#. Ensure all ``MyModel`` references point to the new model location.
#. Run :djadmin:`makemigrations`.

This generates the multiple migration files which do the following:

#. In ``app1``, a migration to rename ``MyModel``'s database table to
   ``app2_mymodel``. This is only if ``MyModel`` does not define a
   ``Meta.db_table`` in ``app2`` that matches the old table name of ``app1``.
   In other words, this is only if the table name is implicit.

#. In ``app2``, a migration to create ``MyModel`` in ``app2`` with the option
   ``managed=False``. At this point, the project has two models pointing to the
   same table.

#. In ``app1``, a migration to mark ``MyModel`` in ``app1`` as
   ``managed=False``. This later makes it possible to remove the old model
   without deleting the table.

#. In ``app2``, a migration to mark ``MyModel`` in ``app2`` as ``managed=True``.

#. In ``app1``, a migration to delete ``MyModel`` from ``app1``.

.. admonition:: Moving models that use multi-table inheritance is not supported

    A model that uses :ref:`multi-table inheritance <multi-table-inheritance>`
    must be moved together with all inherited models.
```

I tried to rewrite the docs for clarity

which reference the field exist. For example, after squashing migrations and
removing the old ones, you should be able to remove the field completely.

.. _moving-model-between-apps:

Moving a model between apps
===========================

> .. versionadded:: 5.2

Reviewer 2: (time flies...)
```suggestion
.. versionadded:: 6.1
```

Django handles the process of transferring models from one app to another. This
involves recognizing the moved model and automatically generating the necessary
migrations.

To move a model ``MyModel`` from ``app1`` to ``app2``:

#. Move the ``MyModel`` class definition from ``app1`` to ``app2``.
#. Ensure all ``MyModel`` references point to the new model location.
#. Run :djadmin:`makemigrations`.

> This generates the multiple migration files which do the following:

Reviewer 2: ```suggestion
If the model has been moved without changes to field definitions,
Django will detect it as a model move. It will ask for your confirmation,
and if confirmed, generate the following migrations:
```

> #. In ``app1``, a migration to rename ``MyModel``'s database table to
>    ``app2_mymodel``. This is only if ``MyModel`` does not define a
>    ``Meta.db_table`` in ``app2`` that matches the old table name of ``app1``.
>    In other words, this is only if the table name is implicit.

Reviewer 2: I find the current phrasing confusing and convoluted.
```suggestion
#. In ``app1``, a migration to rename ``MyModel``'s database table, if
   necessary. By default, the table needs to be renamed from ``app1_mymodel``
   to ``app2_mymodel``.
```
